### PR TITLE
feat(pallet-gear-rpc): Pass gas_allowance multiplier from client

### PR DESCRIPTION
**Pull request by ukint-vs**
Resolves an issue when we have harcoded limit in runtime for runtime api calls.
With these changes we can pass custom `gas_allowance` multiplier for runtime api calls and not worry about future compatibility which requires runtime upgrade.

- runtime api version bump 2
- signature of rpc calls remains the same

todo:

- [x] Check support for older runtimes
- [x] Implement cli parameter for `gas_allowance`
- [x] Fix tests

@gear-tech/dev

### CODE REVIEW

File: pallets/gear/rpc/src/lib.rs [hunk 1]
```diff
@@ -313,34 +444,82 @@ where
...
+            .map_err(|e| map_err(e, "Failed to get gear runtime api version"))?
+            .ok_or_else(|| {
+                CallError::Custom(ErrorObject::owned(
+                    8000,
+                    "Gear runtime api wasn't found in the runtime",
+                    None::<String>,
+                ))
+            })?;
+
+        if api_version < 2 {
```
Comment (shamilsan) [inline pallets/gear/rpc/src/lib.rs:460]: We can avoid repeating `self.run_with_api_copy` here.

```rust
let closure = if api_version < 2 {
    |api| {
                #[allow(deprecated)]
                api.calculate_gas_info_before_version_2(...)
    }
else {
    |api| {
                api.calculate_gas_info(...)
    }
};
self.run_with_api_copy(closure);
```

File: pallets/gear/rpc/src/lib.rs [hunk 2]
```diff
@@ -313,34 +444,82 @@ where
...
+            .client
+            .runtime_api()
+            .api_version::<dyn GearRuntimeApi<Block>>(at_hash)
+            .map_err(|e| map_err(e, "Failed to get gear runtime api version"))?
+            .ok_or_else(|| {
+                CallError::Custom(ErrorObject::owned(
+                    8000,
+                    "Gear runtime api wasn't found in the runtime",
+                    None::<String>,
+                ))
```
Comment (shamilsan) [inline pallets/gear/rpc/src/lib.rs:457]: We could also make `map_err/[in]to_call_error` more general for this case.

File: runtime/common/src/apis.rs [hunk 3]
```diff
@@ -157,8 +157,9 @@ macro_rules! impl_runtime_apis_plus_common {
 					value: u128,
 					allow_other_panics: bool,
 					initial_gas: Option<u64>,
+					gas_allowance: Option<u64>,
 				) -> Result<pallet_gear::GasInfo, Vec<u8>> {
-					Gear::calculate_gas_info(account_id, kind, payload, value, allow_other_panics, initial_gas)
+					Gear::calculate_gas_info(account_id, kind, payload, value, allow_other_panics, initial_gas,gas_allowance)
```
Comment (shamilsan) [inline runtime/common/src/apis.rs:162]:
```suggestion
					Gear::calculate_gas_info(account_id, kind, payload, value, allow_other_panics, initial_gas, gas_allowance)
```

File: runtime/common/src/apis.rs [hunk 4]
```diff
@@ -177,12 +178,13 @@ macro_rules! impl_runtime_apis_plus_common {
 					fn_name: Vec<u8>,
 					wasm: Vec<u8>,
 					argument: Option<Vec<u8>>,
+					gas_allowance: Option<u64>,
 				) -> Result<Vec<u8>, Vec<u8>> {
-					Gear::read_state_using_wasm(program_id, payload, fn_name, wasm, argument)
+					Gear::read_state_using_wasm(program_id, payload, fn_name, wasm, argument,gas_allowance)
```
Comment (shamilsan) [inline runtime/common/src/apis.rs:183]:
```suggestion
					Gear::read_state_using_wasm(program_id, payload, fn_name, wasm, argument, gas_allowance)
```

File: pallets/gear/rpc/src/lib.rs [hunk 5]
```diff
@@ -31,12 +31,13 @@ use jsonrpsee::{
...
 pub use pallet_gear_rpc_runtime_api::GearApi as GearRuntimeApi;
 use pallet_gear_rpc_runtime_api::{GasInfo, HandleKind};
-use sp_api::{ApiError, ApiRef, ProvideRuntimeApi};
+use sp_api::{ApiError, ApiExt, ApiRef, ProvideRuntimeApi};
 use sp_blockchain::HeaderBackend;
 use sp_core::{Bytes, H256};
 use sp_runtime::traits::Block as BlockT;
 use std::sync::Arc;
 
+const GAS_ALLOWANCE: u64 = 600_000_000;
```
[the remark this answers is not in the document]
Comment (ukint-vs) [reply pallets/gear/rpc/src/lib.rs]: removed

File: pallets/gear/rpc/src/lib.rs (hunk 1 above)
Comment (ukint-vs) [reply pallets/gear/rpc/src/lib.rs]: moved into trait

File: node/cli/src/cli.rs [hunk 6]
```diff
@@ -33,6 +33,10 @@ pub struct RunCmd {
     /// The upper limit for the amount of gas a validator can burn in one block.
     #[arg(long)]
     pub max_gas: Option<u64>,
+
+    /// The upper limit for the amount of gas a runtime api can burn in one call.
+    #[arg(long, default_value_t = 64)]
+    pub rpc_gas_allowance_multiplier: u64,
```
Comment (breathx) [inline node/cli/src/cli.rs:39]: Mb?
```suggestion
    pub rpc_calculations_multiplier: u64,
```

Comment (breathx) [reply node/cli/src/cli.rs:39]: You also need different arg for limitations of batch len

File: pallets/gear/src/runtime_api.rs [hunk 7]
```diff
@@ -110,8 +119,9 @@ where
 
         let mut ext_manager = ExtManager::<T>::default();
 
-        GasAllowanceOf::<T>::put(BlockGasLimitOf::<T>::get() * RUNTIME_API_BLOCK_LIMITS_COUNT);
-        QueueProcessingOf::<T>::allow();
+        let _ = Self::update_gas_allowance(
+            allowance_multiplier.unwrap_or(RUNTIME_API_BLOCK_LIMITS_COUNT),
```
Comment (breathx) [inline pallets/gear/src/runtime_api.rs:123]: multiply const by block weight

File: pallets/gear/src/runtime_api.rs [hunk 8]
```diff
@@ -310,28 +321,37 @@ where
...
+            payload_arg,
+            allowance_multiplier,
+        )?);
 
         let block_info = BlockInfo {
             height: Self::block_number().unique_saturated_into(),
             timestamp: <pallet_timestamp::Pallet<T>>::get().unique_saturated_into(),
         };
 
+        let gas_allowance = Self::update_gas_allowance(
```
Comment (breathx) [inline pallets/gear/src/runtime_api.rs:335]: same here
don't mess up or down-lying api with explicit allowance num and passed high level multiplier

File: pallets/gear/src/tests.rs [hunk 9]
```diff
@@ -232,7 +232,7 @@ fn read_big_state() {
 
             assert_succeed(mid);
             let state =
```
Comment (breathx) [inline pallets/gear/src/tests.rs:234]: any test for new functionality?

**Review by breathx — CHANGES_REQUESTED** [review]

File: pallets/gear/src/tests.rs (hunk 9 above)
Comment (breathx) [reply pallets/gear/src/tests.rs:234]: err-case please

**Review by breathx — APPROVED** [review]

**Review by shamilsan — APPROVED** [review]
Very useful update 🆒

**Pull request merged**